Redesign the build to use dependsOn instead of publishLocal.

sjrd opened the pull request:
Previously, the build made heavy use of `publishLocal` to build and publish its artifacts, which were then used by other artifacts with `libraryDependencies`. This caused a lot of time spent in rebuilding and republishing artifacts all the time, and required the use of a special `rebuild` command to clean things up.

In this commit, we turn all those implicit dependencies into actual internal dependencies, using `dependsOn`. This allows sbt to correctly track internal dependencies itself, and rebuild when needed.

The only tricky aspects is that some artifacts must absolutely be exposed in their .jar form to the rest of the build:

* the compiler plugin, because scalac can only load compiler plugins from jars
* `nativelib`, because the linker assumes that it is a jar (it opens it to extract native C code from it)
* some other core libraries, because they manipulate their `products` (e.g., removing generated .class files)

We simply use `exportJars := true` to enforce that.

In order to cancel the `libraryDependencies` added by the `ScalaNativePlugin` on the artifacts built in the build itself, we create a custom `MyScalaNativePlugin` on top of it, which removes them. Instead of relying on those automatic `libraryDependencies`, we must explicitly tell all the required dependencies (and only those) as `dependsOn` clauses.

---

The above changes mean that most sbt commands, like `tests/test`, will not test the artifacts in their published form anymore. Instead, they will test the jars produced within the build. The published artifact are still tested through the scripted tests of the sbt plugin, so we are not losing any test coverage in the process.

---

In the process, we cleaned up several stale or dubious internal dependencies. This eventually led to rethink what classpaths are passed down to the `tools/test` task, so that was cleaned up as well.

---

The `rebuild` command is not necessary anymore. `clean` should be used instead.

The `dirty-rebuild` command is not necessary anymore. Simply do *nothing* instead.

The `cleanCache` and `cleanLocal` commands, provided by sbt-dirty-money, and which used to be used by `rebuild`, are not necessary anymore either. Therefore, we have completely removed sbt-dirty-money. If a developer still wants to use these commands for their own purposes, they can globally install sbt-dirty-money, as suggested in their readme.

sjrd commented:
@LeeTibbert @ekrich Can you validate that this does not break your typical workflow (other than removing `rebuild` and `dirty-rebuild` from your muscle memory).

LeeTibbert commented:
Thank you for the opportunity to study this work. I have been beating_on/exercising_it now
for 5 hours or more.   It is a quantum leap forward & amazing!  At the highest level, I believe
it should move forward.  

I encountered one large issue on my system, which I will add in
a separate comment below.  I also a medium-small and several small
issues. Again, I will add them as comments as I can create the time.

It was nice to see the Contributor's Guide updated to match.  That helped me adjust to the
new environment.

The new build reduced my initial build time after I reached the sbt prompt from about
5 minutes to just over 3.  Yeah!  Some of the savings may have come from what I think
is a slightly increased time to reach the sbt prompt (understandable because of new .scala
build file.)  I'll take the trade-off any day.

This build addressed my concern that new devos be left in a stable state after they
clone & reach the sbt prompt.  No more hosing them with a missing nscplugin.
The "reasonable", guessable, or easily gleaned from build.sbt, commands, 
"sandbox/run" and "test-all" all work.

I did "black-box" testing playing the role of a new devo, 
read the new doc, and only spot checked the code itself. All good.
This is a holiday long weekend in America, so it may be a few
days before I start point review.  Please do not let me hold anything up.

LeeTibbert commented:
The major problem that I have and that I propose be fixed is that when I followed the
directions in the new Contributor's Guide and ran "sandbox/run" immediately after
a clean clone (with clean .ivy cache, local, & coursier cache), I got
```
[error] Could not find nativelib on classpath.
[error] (sandbox / Compile / nativeConfig) Could not find nativelib on classpath.
```
I have been following the progress of this PR, so I know that the Travis CI build worked.
Rats!

Turned out to be a problem/bug in `tools/src/main/scala/scala/scalanative/build/Discover.scala`

/** Find nativelib jar on the classpath. */
  def nativelib(classpath: Seq[Path]): Option[Path] =
    classpath.find { path =>
      val absolute = path.toAbsolutePath.toString
      absolute.contains("scala-native") && absolute.contains("nativelib")    
    }
```
Turn out the directory path to the newly build nativelib.jar does not contain "scala-native"
so the test was failing.  An obviously present jar was not found.   I did not track down
the naming in Travis CI to see why that worked.

When some blatant hacker too ashamed to admit to it changed the test, `sandbox/run`
and `test-runtime` both work.  
```
/** Find nativelib jar on the classpath. */
  def nativelib(classpath: Seq[Path]): Option[Path] =
    classpath.find { path =>
      val absolute = path.toAbsolutePath.toString
     // find the new .jar
   
      (absolute.contains("nativelib") &&
      absolute.contains("target") &&
      absolute.contains("scala-2.11")) ||
      (absolute.contains("scala-native") && absolute.contains("nativelib"))  // find artifact jar
    }
```
Parts of scripted-tests failed without retaining the part to find the artifact jar.
Again, the above code is crufty. It is just to illustrate a workaround.

Getting `test-runtime` to work was a, for me, achievement. I have not
done timing studies to see if that command executes any faster.

I also have not counted the number of publishLocals to see if the prior 2 publishLocals
of nscplugin in `test-all` are gone. Given that it is, I believe, no longer published in 
what-replaces-rebuild, one publishLocal might be acceptable.

LeeTibbert commented:
This PR removes the dirty-money plugin so the `cleanCache` & `cleanLocal` commands
are no longer available.  I suggest listing these along side of `rebuild` as no longer available.

 I think scripted-tests and possibly other tests still write into the local .ivy2 cache. Being able
to clean of SN artifacts, especially -SNAPSHOT gives me a sense of comfort and is arguably
a feature which should be in the SN build.

I personally can restore dirty-money in a project/local.sbt.  After this PR, a _brief_ section
about cleaning caches in the Contributor's Guide would be worthwhile.  I am willing to
write a draft  and circulate it for comments.  I am trying to work out things that might
well trip up new devos.

I am still getting thrashed since sbt 1.3.8 by an inability to simply clean the Coursier build,
 inside or outside of SN build.sbt. To my thinking a cleanAll alias which does the sbt clean,
 cleanCache, cleanLocal and whatever is need to clean Coursier, especially -SNAPSHOTs.
 I am aware of COURSIER_TTL=0s but it is a pain to remember in each new session.

LeeTibbert commented:
`sbt publishLocal` appears to publish one more artifact into the local repository.
There are now 18 artifacts.  Mainline publishes 17.

I believe the extra artifact is the (missing) root project for SN. Because the build.sbt
has no explicit root project, the artifact is given the name of the project base directory.
I believe I have seen this discussed in some other project.

I am reporting this because I was asked to keep an eye out for changed behavior.

I think, but have not tried, that the solution is to have an explicit root project which
sets its publish & publishLocal values to false. 

I believe that some unnecessary `stuff` is being put into the baseDirectory/target.
I do not remember what used to be there. I think whatever it was, it was also pretty useless.
Now there seems to be more.

Sorry to be vague, I've traveled over a number of hard issues today.

LeeTibbert commented:
I can confirm that 15 artifacts get publishedLocal and once apiece.

There is one small change in behavior which should probably go into the README for
this PR.  From study of build.sbt, this new behavior is intentional.  There is a certain
logic to testing .jars before putting them into artifacts, then testing the artifacts. 
The new behavior is just a change in feed.

Previously the `tests/test` would use previously published artifacts.  If those
artifacts did not exist, it reported the error and halted.   Now `tests/test` and those
aliases which use it, test against the .jar files.  

`sbtScalaNative/scripted` publishes the artifacts and tests against them. 

A discussion for the future: I am not certain that each of the artifacts are indeed tested.

sjrd commented:
Thanks a lot @LeeTibbert for your detailed comments.

> The major problem that I have and that I propose be fixed is that when I followed the
> directions in the new Contributor's Guide and ran "sandbox/run" immediately after
> a clean clone (with clean .ivy cache, local, & coursier cache), I got
> 
> ```
> [error] Could not find nativelib on classpath.
> [error] (sandbox / Compile / nativeConfig) Could not find nativelib on classpath.
> ```
> 
> I have been following the progress of this PR, so I know that the Travis CI build worked.
> Rats!
> 
> Turned out to be a problem/bug in `tools/src/main/scala/scala/scalanative/build/Discover.scala`
> 
> /** Find nativelib jar on the classpath. */
> def nativelib(classpath: Seq[Path]): Option[Path] =
> classpath.find { path =>
> val absolute = path.toAbsolutePath.toString
> absolute.contains("scala-native") && absolute.contains("nativelib")
> }
> 
> ```
> Turn out the directory path to the newly build nativelib.jar does not contain "scala-native"
> so the test was failing.  An obviously present jar was not found.   I did not track down
> the naming in Travis CI to see why that worked.

Ouch, that's bad! It works on my system and on Travis because the repo itself is cloned in a directory called `scala-native/`. It's fortunate that yours is not, so that you could catch this issue.

I am trying to find a solution for this.

> This PR removes the dirty-money plugin so the `cleanCache` & `cleanLocal` commands
> are no longer available. I suggest listing these along side of `rebuild` as no longer available.

Good point. I will add this to the commit message.

> I think scripted-tests and possibly other tests still write into the local .ivy2 cache. Being able
> to clean of SN artifacts, especially -SNAPSHOT gives me a sense of comfort and is arguably
> a feature which should be in the SN build.

It doesn't need to be in the SN build itself because the build and typical workflows with it have no reason left to run those commands. The scripted tests still do publishLocal`, but they will always publish everything needed for the tests, so that will simply overwrite any previously published artifacts. There is no need to manually clean the `/local/` directory anymore.

> I personally can restore dirty-money in a project/local.sbt. After this PR, a _brief_ section
> about cleaning caches in the Contributor's Guide would be worthwhile. I am willing to
> write a draft and circulate it for comments. I am trying to work out things that might
> well trip up new devos.

Again, there is no reason that you should ever need to run `cleanLocal` or `cleanCache` ever again, at least not because of the build in Scala Native. If you still want it for you, you can put in your global sbt config, but then that's already part of the sbt-dirty-money Readme. Since it's there, and not needed by the Scala Native build, there's no reason to put that in the Scala Native contributing guide anymore.

> There is one small change in behavior which should probably go into the README for
> this PR. From study of build.sbt, this new behavior is intentional. There is a certain
> logic to testing .jars before putting them into artifacts, then testing the artifacts.
> The new behavior is just a change in feed.
> 
> Previously the `tests/test` would use previously published artifacts. If those
> artifacts did not exist, it reported the error and halted. Now `tests/test` and those
> aliases which use it, test against the .jar files.
> 
> `sbtScalaNative/scripted` publishes the artifacts and tests against them.
> 
> A discussion for the future: I am not certain that each of the artifacts are indeed tested.

This is indeed intentional. We rely on the scripted tests to make sure that the artifacts can be published and used at all. But the groundwork of testing all the intricate behavior of the compiler, libraries and tools is done within the build, directly on the .jars. This is the same strategy that we apply in Scala.js, and we never had any issue of wrong artifacts when the jars passed the tests. So it should be fine.

sjrd commented:
@LeeTibbert I have updated the commit message with more info on the tests-are-on-jars story, and on what happened to `cleanCache` and `cleanLocal`.

I have also fixed the "Could not find nativelib" issue, in the second commit.

ekrich commented:
@sjrd I'm sorry I haven't been able to take a look at this PR. I started a small remodel project of our laundry room on Friday and I am trying to get it done this long holiday weekend. I have read along and it sounds good though.

LeeTibbert commented on docs/contrib/build.rst:
```diff
@@ -8,30 +8,32 @@ This section gives some basic information and tips about the build system. The
+- ``tools/test`` -- run the unit tests of the tools, aka the linker
+- ``sbtScalaNative/scripted`` -- run the integration tests of the sbt plugin
+  (this takes a while)
+- ``clean`` -- clean all compiled artifacts
```
also cleans deletes cached scala sources

sjrd replied:
done.

LeeTibbert commented on docs/contrib/build.rst:
```diff
@@ -111,75 +114,62 @@ The following shows how to set ``nativeGC`` on all the projects.
+    - ``auxlib``
 
-    - `scalalib`
+    - ``scalalib``
```
Since these are not in alphabetical order, or in another obvious order,
it may be worthwhile to say that projects later in the list depend on those
earlier in the list.

sjrd replied:
Done.

LeeTibbert commented:
Both re-wordings read well.  Thank you for taking the time.  I think they will help both
 present & future contributors.

sjrd commented:
> The documentation is very good as well. I think there are some small changes needed on the following page:
> https://github.com/scala-native/scala-native/blob/master/docs/contrib/compiler.rst

Indeed, thanks for spotting that! I have updated that document.

> So generally, testing can be done in the `sandbox` project which I have never used as described above.
> 
> The second approach is to `publishLocal` and then use a separate project pointing at the `-SNAPSHOT` built. I tested `publishLocal` and used a separate project and that went great. It might be worth cross referencing the `compiler.rst` page as well as adding some text similar to the above about `publishLocal` in `build.rst` for newer contributors.

I added some information about that in `build.rst`.

> I think it might be useful to look at the changes made in this commit - [a93677e](https://github.com/scala-native/scala-native/commit/a93677e93c8d52049b19563fdd229fbaa37b2656) in relation to this PR - #1637
> 
> I know you started a review on this and we haven't gone back to look at it again but I think some of the changes here using ModuleID and having some case classes in the `build.Discover` object are quite useful for a more general build. At a minimum, the PR is a really good start for this proposal - #1352

I think the changes here are going to simplify #1637. Indeed, they highlight that what the `Build` wants is just the .jar for native code, next to the classpath, and that it really does not want to be looking at the *names* of the jar files, which are irrelevant. The sbt plugin is the one responsible for identifying the right jar, instead of `Discover`. Consequently, in #1637 the concepts of `LibID` and `Nativelib` won't be necessary anymore.

> What is the use for this or'ed piece of code?
> https://github.com/scala-native/scala-native/pull/1810/files#diff-9f8176d7a89fed51f9b5c82605327b36R107-R108

Hum ... isn't the comment 3 lines above clear enough?

ekrich commented:
@sjrd I did read the comment about the cross-version suffix but was wondering if that is something you encountered now or envision encountering in the future?

My thinking was that the Tools API should do most of the work internally so people integrating like the sbt plugin, Mill, or coursier wouldn't have to. Also, the C/C++ code gets copied into a `nativelib` dir so passing the information in came in handy to put the code into directories that are unique based on the organization and name rather than hard coded. I assume that we will always have the _ModuleId_ parts coming from any build tool.

When I looked at it this morning I started to think that the following code should be in `tools` as well.
https://github.com/ekrich/scala-native/blob/topic/fix-657/sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala#L131-L138

And line 144 as well.

sjrd commented:
> I did read the comment about the cross-version suffix but was wondering if that is something you encountered now or envision encountering in the future?

It happens now. It's `nativelib` while in the build, but `nativelib_native0.4.0-SNAPSHOT_2.11` once outside the build (for example in scripted tests).

> My thinking was that the Tools API should do most of the work internally so people integrating like the sbt plugin, Mill, or coursier wouldn't have to. Also, the C/C++ code gets copied into a nativelib dir so passing the information in came in handy to put the code into directories that are unique based on the organization and name rather than hard coded. I assume that we will always have the ModuleId parts coming from any build tool.

Not all jars even have a module ID. For example jars in `lib/` are on the classpath, but do not have a module ID. Module IDs are a concept of *resolution of dependencies*, not a concept of classpaths. As such, it is in fact meaningless for the tools to try and apply module ID-related treatment on its classpath entries.

Even in the sbt plugin, I didn't want to do that. I wanted to go straight from the update report. But that proved very tedious for technical reasons, and I noticed that I could in fact retrieve the module IDs from `fullClasspath`, I jumped on it. But that's awkward in itself, and while I can tolerate awkward in the sbt plugin, we shouldn't build awkward in the tools.

And even if we did that, we should definitely pass along the module IDs from the dependency resolution to the tools (like Attributes do to pass them to the gym classpath) together than rely on the paths of the jars, like was done before this PR, because the paths of jars are not specified whatsoever, and they most definitely do not always contain the organization name.

> Edit: Not to belabor the point but couldn't [...] be [...]?

No, because that would also match `nativeliboops`.

ekrich commented:
Ok sounds good. I obviously don't know all the in and outs so that is why I was asking the questions. I am sure we can find better ways to implement to tools going forward as well.
All my concerns have been answered so LGTM.

sjrd commented:
Thanks a lot @errikos @LeeTibbert @ekrich for your reviews :)
